Mathlib/Algebra/Category/Ring/Colimits.lean

```diff
@@ -0,0 +1,397 @@
+  | mul_one : ∀ x, Relation (mul x one) x
+  | add_left_neg : ∀ x, Relation (add (neg x) x) zero
+  | add_comm : ∀ x y, Relation (add x y) (add y x)
+  | mul_comm : ∀ x y, Relation (mul x y) (mul y x)
+  | add_assoc : ∀ x y z, Relation (add (add x y) z) (add x (add y z))
+  | mul_assoc : ∀ x y z, Relation (mul (mul x y) z) (mul x (mul y z))
+  | left_distrib : ∀ x y z, Relation (mul x (add y z)) (add (mul x y) (mul x z))
+  | right_distrib : ∀ x y z, Relation (mul (add x y) z) (add (mul x z) (mul y z))
+#align CommRing.colimits.Relation CommRingCat.Colimits.Relation
+
+/-- The setoid corresponding to commutative expressions modulo monoid Relations and identifications.
+-/
+def colimitSetoid : Setoid (Prequotient F) where
+  r := Relation F
+  iseqv := ⟨Relation.refl, Relation.symm _ _, Relation.trans _ _ _⟩
+#align CommRing.colimits.colimit_setoid CommRingCat.Colimits.colimitSetoid
+
+attribute [instance] colimitSetoid
+
+/-- The underlying type of the colimit of a diagram in `CommRing`.
+-/
+def ColimitType : Type v :=
+  Quotient (colimitSetoid F)
+#align CommRing.colimits.colimit_type CommRingCat.Colimits.ColimitType
+
+-- Porting note : failed to derive `Inhabited` instance
+instance InhabitedColimitType : Inhabited <| ColimitType F where
+  default := Quot.mk _ zero
+
+instance ColimitType.AddGroup : AddGroup (ColimitType F) where
+  zero := Quot.mk _ zero
+  neg := by
+    fapply @Quot.lift
+    · intro x
+      exact Quot.mk _ (neg x)
+    · intro x x' r
+      apply Quot.sound
+      exact Relation.neg_1 _ _ r
+  add := by
+    fapply @Quot.lift _ _ (ColimitType F → ColimitType F)
+    · intro x
+      fapply @Quot.lift
+      · intro y
+        exact Quot.mk _ (add x y)
+      · intro y y' r
+        apply Quot.sound
+        exact Relation.add_2 _ _ _ r
+    · intro x x' r
+      funext y
+      refine Quot.inductionOn y ?_
+      exact fun _ => Quot.sound (Relation.add_1 _ _ _ r)
+  zero_add x := Quot.inductionOn x fun _ => Quot.sound (Relation.zero_add _)
+  add_zero x := Quot.inductionOn x fun _ => Quot.sound (Relation.add_zero _)
+  add_left_neg := Quot.ind fun x => by
+    simp only [(. + .)]
+    exact Quot.sound (Relation.add_left_neg x)
+  add_assoc x y z := by
+    refine Quot.induction_on₃ x y z (fun a b c => ?_)
+    simp only [(. + .)]
+    apply Quot.sound
+    apply Relation.add_assoc
+
+instance ColimitType.AddGroupWithOne : AddGroupWithOne (ColimitType F) :=
+  { ColimitType.AddGroup F with one := Quot.mk _ one }
+
+-- Porting note : In Lean 4, `Ring` needs a proof of `mul_zero` and `zero_mul`, these are not
+-- entirely obvious from `Relation F`. So I first prove that `ColimitType F` is such that
+-- `a + b = a + c → b = c`. Then prove that `a * 0 + a * 0 = a * 0 + 0` hence `a * 0 = 0`.
+-- In `mathlib3`, `ring` axioms do not include `mul_zero` and `zero_mul`, so this is not necessary.
+-- This might not be the best solution. For example, one can add addition axioms to `Relation F`.
```

Comment:
I think adding the additional axioms to `Relation F` is the best way. There are already superfluous axioms there such as `add_comm`.